scvi/data/fields/_protein.py

```diff
@@ -110,3 +95,79 @@ def transfer_field(
             transfer_state_registry[self.PROTEIN_BATCH_MASK] = batch_mask
 
         return transfer_state_registry
+
+
+class ProteinObsmField(ProteinFieldMixin, ObsmField):
+    """
+    An AnnDataField for an protein data stored in an .obsm field of an AnnData object.
+
+    For usage with the TotalVI model. Computes an additional mask which indicates
+    where batches are missing protein data.
+
+    Parameters
+    ----------
+    registry_key
+        Key to register field under in data registry.
+    obsm_key
+        Key to access the field in the AnnData .obsm mapping.
+    use_batch_mask
+        If True, computes a batch mask over the data for missing protein data.
+        Requires ``batch_key`` to be not None.
+    batch_key
+        Key corresponding to the .obs field where batch indices are stored.
+        Used for computing a batch mask over the data for missing protein data.
+    colnames_uns_key
+        Key to access column names corresponding to each column of the .obsm field in
+        the AnnData .uns mapping. Only used when .obsm data is a np.ndarray, not a pd.DataFrame.
+    is_count_data
+        If True, checks if the data are counts during validation.
+    correct_data_format
+        If True, checks and corrects that the AnnData field is C_CONTIGUOUS and csr
+        if it is dense numpy or sparse respectively.
+    """
+
+
+class ProteinLayerField(ProteinFieldMixin, LayerField):
+    """
+    An AnnDataField for an protein data stored in a layer field of an AnnData object.
+
+    For usage with the TotalVI model. Computes an additional mask which indicates
+    where batches are missing protein data.
+
+    Parameters
+    ----------
+    registry_key
+        Key to register field under in data registry.
+    layer
+        Key to access the field in the AnnData layers mapping. If None, uses the data in .X.
+    use_batch_mask
+        If True, computes a batch mask over the data for missing protein data.
+        Requires ``batch_key`` to be not None.
+    batch_key
+        Key corresponding to the .obs field where batch indices are stored.
+        Used for computing a batch mask over the data for missing protein data.
+    is_count_data
+        If True, checks if the data are counts during validation.
+    correct_data_format
+        If True, checks and corrects that the AnnData field is C_CONTIGUOUS and csr
+        if it is dense numpy or sparse respectively.
+    """
+
+
+def copy_over_batch_attr(self, mdata: MuData):
+    # Assign self.batch_field if not yet assigned to MuDataWrapped field.
+    # Then, reassign self.adata_field.batch_field to the batch AnnDataField.
+    if isinstance(self.adata_field.batch_field, BaseMuDataWrapperClass):
+        self.batch_field = self.adata_field.batch_field
+        self.adata_field.batch_field = self.batch_field.adata_field
+
```

Comment:
docstring so we don't forget what this does

Reply:
also @jjhong922 can you add docstrings to the whole wrapper stuff? The preregister fn especially